### ISSUE TITLE
Crash when orientation changes using Platform Views in a ListView

### ISSUE THREAD

**Issue by raycardillo-VerizonMedia** [flutter/flutter#45443]
I see that some **Platform View crashes** were resolved in the engine recently, but when I was testing that they were all resolved properly, I noticed this problem on another one of our screens. The difference with this screen is that multiple Platform Views are in a ListView and the crash happens only after several orientation changes. On the simulator we can see corruption/distortion (but no crash). On device it crashes after about 4 rotations. Sample code attached.

## Steps to Reproduce

See sample code:
https://github.com/raycardillo-VerizonMedia/flutter-rotation-bug

See instructions in the **README** but it's not difficult to reproduce at all. Just run on device and rotate the device several times (usually 4 times on my iPhone X and iPhone 11 Pro Max both running the latest available iOS 13).

**Target Platform:** iOS (maybe also on Android but I cannot test WebView in `v1.12`)
**Target OS version/browser:** iOS 13
**Devices:** iPhone X, iPhone 11 Pro Max

## Logs

```
no relevant errors in verbose output
```

```
no flutter analyze issues
```

### `flutter doctor -v`
```
[✓] Flutter (Channel master, v1.12.10-pre.12, on Mac OS X 10.14.6 18G1012, locale en-US)
    • Flutter version 1.12.10-pre.12 at /Users/rcardillo15/Library/flutter
    • Framework revision b6e92003c8 (89 minutes ago), 2019-11-22 15:02:20 -0800
    • Engine revision ad14574949
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /android-sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /android-sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.1)
    • Xcode at /Applications/Xcode-11.1.app/Contents/Developer
    • Xcode 11.1, Build version 11A1027
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (1 available)
    • Mr. Mighty • 00008030-0005644A2644802E • ios • iOS 13.2.3

• No issues found!
```

**Comment by raycardillo-VerizonMedia** [comment]
See attached for an example of the problems that happen in the simulator. In this case, it was the AppBar, but sometimes the titles are in the wrong places. On device, this crashes after about 4 rotations.

![flutter-rotate-bug](https://user-images.githubusercontent.com/58046798/69470902-04696f80-0d68-11ea-91b8-973dbba4594b.gif)

**Comment by raycardillo-VerizonMedia** [comment]
@gaaclarke - @eseidelGoogle said he thinks you're working on Metal? If not, can you tag whoever might be interested in this issue? I implemented a quick example (see above GitHub repo) with the hope of getting more immediate attention on this issue. This is the new issue I was trying to report in #32453. I don't think these issues are related, but I found it while testing other crashes that were happening when multiple Platform Views are displayed and the device is rotated (e.g. #34330).

Also, if possible, this case should be tested as well when implementing #31139 so we don't regress on this behavior in the future.

**Comment by gaaclarke** [comment]
Thanks @raycardillo-VerizonMedia, I'll take a look.   Can you please include a stacktrace for the crash?

**Comment by gaaclarke** [comment]
I was able to reproduce this.  I had problems reproducing it on my iPhone SE on iOS 11.3.  On my iPhone 8 on iOS 13 it wasn't hard.

Here is the stacktrace:
```
Thread 1 Queue : com.apple.main-thread (serial)
#0	0x000000010151e340 in sk_sp<SkSurface>::get() const at /Users/aaclarke/dev/engine/src/third_party/skia/include/core/SkRefCnt.h:298
#1	0x000000010151e30c in sk_sp<SkSurface>::operator bool() const at /Users/aaclarke/dev/engine/src/third_party/skia/include/core/SkRefCnt.h:296
#2	0x000000010151c148 in bool operator!=<SkSurface>(sk_sp<SkSurface> const&, std::nullptr_t) at /Users/aaclarke/dev/engine/src/third_party/skia/include/core/SkRefCnt.h:352
#3	0x0000000101783ce8 in flutter::SurfaceFrame::SkiaCanvas() at /Users/aaclarke/dev/engine/src/flutter/shell/common/surface.cc:36
#4	0x00000001014aea6c in flutter::FlutterPlatformViewsController::SubmitFrame(GrContext*, std::__1::shared_ptr<flutter::IOSGLContext>) at /Users/aaclarke/dev/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:373
#5	0x000000010151a33c in flutter::IOSSurfaceGL::SubmitFrame(GrContext*) at /Users/aaclarke/dev/engine/src/flutter/shell/platform/darwin/ios/ios_surface_gl.mm:148
#6	0x000000010151a3fc in non-virtual thunk to flutter::IOSSurfaceGL::SubmitFrame(GrContext*) ()
#7	0x00000001016f789c in flutter::Rasterizer::DrawToSurface(flutter::LayerTree&) at /Users/aaclarke/dev/engine/src/flutter/shell/common/rasterizer.cc:318
#8	0x00000001016f88c4 in flutter::Rasterizer::DoDraw(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >) at /Users/aaclarke/dev/engine/src/flutter/shell/common/rasterizer.cc:222
#9	0x00000001016fd474 in flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0::operator()(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >) const at /Users/aaclarke/dev/engine/src/flutter/shell/common/rasterizer.cc:123
#10	0x00000001016fd3f8 in decltype(std::__1::forward<flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0&>(fp)(std::__1::forward<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(fp0))) std::__1::__invoke<flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) at /Users/aaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4350
#11	0x00000001016fd370 in void std::__1::__invoke_void_return_wrapper<void>::__call<flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) at /Users/aaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/__functional_base:349
#12	0x00000001016fbcf8 in std::__1::__function::__func<flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0, std::__1::allocator<flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0>, void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>::operator()(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) at /Users/aaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1572
#13	0x00000001016fd8a0 in std::__1::function<void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>::operator()(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >) const at /Users/aaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1923
#14	0x00000001016f7fb0 in flutter::Pipeline<flutter::LayerTree>::Consume(std::__1::function<void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)> const&) at /Users/aaclarke/dev/engine/src/flutter/shell/common/pipeline.h:154
#15	0x00000001016f7ba8 in flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >) at /Users/aaclarke/dev/engine/src/flutter/shell/common/rasterizer.cc:126
#16	0x000000010176983c in flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29::operator()() const at /Users/aaclarke/dev/engine/src/flutter/shell/common/shell.cc:913
#17	0x0000000101769798 in decltype(std::__1::forward<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29&>(fp)()) std::__1::__invoke<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29&>(flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29&) at /Users/aaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4350
#18	0x000000010176974c in void std::__1::__invoke_void_return_wrapper<void>::__call<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29&>(flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29&) at /Users/aaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/__functional_base:349
#19	0x00000001017682a0 in std::__1::__function::__func<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29, std::__1::allocator<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29>, void ()>::operator()() at /Users/aaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1572
#20	0x000000010156cc0c in std::__1::function<void ()>::operator()() const at /Users/aaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1923
#21	0x000000010157fcec in fml::MessageLoopImpl::FlushTasks(fml::FlushType) at /Users/aaclarke/dev/engine/src/flutter/fml/message_loop_impl.cc:123
#22	0x000000010157fb20 in fml::MessageLoopImpl::RunExpiredTasksNow() at /Users/aaclarke/dev/engine/src/flutter/fml/message_loop_impl.cc:133
#23	0x000000010159c228 in fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) at /Users/aaclarke/dev/engine/src/flutter/fml/platform/darwin/message_loop_darwin.mm:75
#24	0x00000001a797ae1c in __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ ()
#25	0x00000001a797ab58 in __CFRunLoopDoTimer ()
#26	0x00000001a797a228 in __CFRunLoopDoTimers ()
#27	0x00000001a7975364 in __CFRunLoopRun ()
#28	0x00000001a79748bc in CFRunLoopRunSpecific ()
#29	0x00000001b17e0328 in GSEventRunModal ()
#30	0x00000001aba0a6d4 in UIApplicationMain ()
#31	0x0000000100f3a11c in main at /Users/aaclarke/dev/flutter-rotation-bug/ios/Runner/main.m:7
#32	0x00000001a77ff460 in start ()
```

It is a bit surprising to see `flutter::Rasterizer::DoDraw` showing up in the stacktrace for the main thread.

**Comment by gaaclarke** [comment]
The following patch removes the crash but leaves the app not drawing the menu bar:
```diff
--- a/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm
+++ b/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm
@@ -370,10 +370,12 @@ bool FlutterPlatformViewsController::SubmitFrame(GrContext* gr_context,
   for (int64_t view_id : composition_order_) {
     EnsureOverlayInitialized(view_id, gl_context, gr_context);
     auto frame = overlays_[view_id]->surface->AcquireFrame(frame_size_);
-    SkCanvas* canvas = frame->SkiaCanvas();
-    canvas->drawPicture(picture_recorders_[view_id]->finishRecordingAsPicture());
-    canvas->flush();
-    did_submit &= frame->Submit();
+    if (frame) {
+      SkCanvas* canvas = frame->SkiaCanvas();
+      canvas->drawPicture(picture_recorders_[view_id]->finishRecordingAsPicture());
+      canvas->flush();
+      did_submit &= frame->Submit();
+    }
   }
   picture_recorders_.clear();
   if (composition_order_ == active_composition_order_) {
```

That will result in the spamming of the "Could not make the context current to acquire the frame" message.  I tracked that down to what is causing it, it is because this call is returning false: https://github.com/gaaclarke/engine/blob/c82556bae4be17e46958b7c96d1f1c6a8c99fc4f/shell/platform/darwin/ios/ios_gl_render_target.mm#L120:L120

**Comment by gaaclarke** [comment]
`IOSRenderTarget::UpdateStorageSizeIfNecessary` returns false because the `layer_` is of an invalid size, like (size_width = 3670, size_height = 0).

Here is the po of the `layer_`:
```
(lldb) po layer_.get()
<CAEAGLLayer:0x280f801c0; position = CGPoint (917.5 0); bounds = CGRect (0 0; 1835 0); delegate = <FlutterOverlayView: 0x107f03f20; frame = (0 0; 1835 0); autoresize = W+H; userInteractionEnabled = NO; layer = <CAEAGLLayer: 0x280f801c0>>; contents = <CAImageQueue 0x2833a8500>; allowsDisplayCompositing = YES; drawableProperties = {
    EAGLDrawablePropertyColorFormat = EAGLColorFormat8888;
    EAGLDrawablePropertyRetained = 0;
}; presentsWithTransaction = 1; rasterizationScale = 2; contentsScale = 2>
```

And the po of the FlutterOverlayView:
```
<FlutterOverlayView: 0x107f03f20; frame = (0 0; 1835 0); autoresize = W+H; userInteractionEnabled = NO; layer = <CAEAGLLayer: 0x280f801c0>>
```

**Comment by raycardillo-VerizonMedia** [comment]
@gaaclarke Thanks for the quick response. Do you know if the fix will also correct the distortion I'm seeing in the simulator? I'm suspecting that it is only going to fix the crash on device. If so, we will have to pursue one problem at a time, but I was just curious because it's seems like there might be more than one issue at play here.

**Comment by gaaclarke** [comment]
@raycardillo-VerizonMedia It should fix the distortion problem since the frame of the views was out of whack.

**Comment by dnfield** [comment]
To be clear - if this is happening on iOS, it's not related to #32453, which is Android specific.

**Comment by raycardillo-VerizonMedia** [comment]
@dnfield - I wasn't sure when I first opened this because I was unable to test on Android. But given that the crash is iOS specific, and that is also causing the distortion in simulator, then it would appear that this issue is iOS specific. (At least until the root cause of the crash is fixed and I can test again on both platforms.) ;)

**Comment by gaaclarke** [comment]
Closed, https://github.com/flutter/engine/pull/14018 merged.
